webpack.config.js

```diff
@@ -0,0 +1,234 @@
+  const browserRender = (env.indexOf('browser') >= 0);
+  console.log(`running webpack.config.js: isProd=${isProd}; browserRender=${browserRender};`);
+
+  /**
+   * __dirname is changed after webpack-ed to another directory
+   * so process.cwd() is used instead to determine the correct base directory
+   */
+  const __DIR__ = process.cwd();
+
+  const PATHS = {
+    app: path.resolve(__DIR__, 'app'),
+    assets: path.resolve(__DIR__, 'public', 'assets'),
+    compiled: path.resolve(__DIR__, 'compiled'),
+    public: '/assets/', // use absolute path for css-loader?
+    modules: path.resolve(__DIR__, 'node_modules')
+  };
+
+  const externals = fs.readdirSync('node_modules')
+    .filter(x => ['.bin'].indexOf(x) === -1)
+    .reduce((acc, cur) => Object.assign(acc, { [cur]: 'commonjs ' + cur }), {});
+
+  const urlLoader = (limit = 10000) => ({
+    test: /\.(png|jpg|jpeg|gif|svg|woff|woff2)$/,
+    loader: 'url-loader',
+    query: { name: '[hash].[ext]', limit }, // name for file-loader
+    include: PATHS.app
+  });
+
+  const babelLoader = (production = false, browser = false) => {
+    const obj = {
+      test: /\.js$|\.jsx$/,
+      loader: 'babel-loader',
+      query: {
+        presets: ['es2015', 'react', 'stage-0'],
+        plugins: ['transform-decorators-legacy']
+      },
+      exclude: PATHS.modules
+    };
+    if (production) {
+      obj.query.plugins.push(
+        'transform-react-remove-prop-types',
+        'transform-react-constant-elements',
+        'transform-react-inline-elements'
+      );
+    } else if (browser) {
+      obj.query.presets.unshift('react-hmre');
+    }
+    return obj;
+  };
+
+  const cssLoader = (production = false, browser = false) => {
+    /*
+     * https://github.com/webpack/css-loader/issues/59 - use css-loader/locals on server-side rendering
+     * :local(.className) used to declare className in the local scope,
+     * local identifiers are exported by css-loader
+     *
+     * modules enable CSS Modules spec https://github.com/css-modules/css-modules
+     *
+     * importLoaders (int): That many loaders after the css-loader are applied to @import-ed resources.
+     *
+     */
+    const queryModules = 'modules';
+    const queryImportLoaders = 'importLoaders=1';
+    const queryLocal = 'localIdentName=[name]__[local]___[hash:base64:5]';
+
+    const queryStrExtractTextPlugin = '?' + queryModules + '&' + queryImportLoaders;
+    const queryStr = queryStrExtractTextPlugin + '&' + queryLocal;
+
+    const serverSide = 'css-loader/locals' + queryStr + '!postcss-loader';
+    const browserSide = 'style-loader!css-loader' + queryStr + '!postcss-loader';
+
+    const myLoader = browser ? browserSide : serverSide;
+
+    const prodBrowserRenderLoader = ExtractTextPlugin.extract({
+      fallbackLoader: 'style-loader',
+      loader: 'css-loader' + queryStrExtractTextPlugin + '!postcss-loader'
+    });
+    const loader = production && browser ? prodBrowserRenderLoader : myLoader;
```

Comment:
Implemented this on a repo based on reactGo over the weekend. One issue I had was that in production the server rendered the dev css module classnames (ie: '.class-name_3efase3') but the css asset was exported using the prodBrowserRenderLoader and rendered the hashed classnames (ie:'.3efase3') so they did not match. Easy way to repo is to run in prod mode and turn off javascript in your browser.

The client would re-render and visually correct it almost instantly (much like it does in dev) but I would expect the server to render the correct classnames.

tl:dr
This should likely be `const loader = production ? prodBrowserRenderLoader : myLoader;`